yapapi/contrib/strategy/rep_a1.py

```diff
@@ -0,0 +1,107 @@
+
+    #################
+    #   OFFER SCORING
+    async def score_offer(self, offer: "OfferProposal") -> float:
+        offer_score = await super().score_offer(offer)
+        provider_score = await get_provider_score(offer.issuer)
+        combined_score = self._final_score(offer_score, provider_score)
+
+        provider_name = offer._proposal.proposal.properties['golem.node.id.name']
+        logger.info(
+            "Scored %s -  base: %s, provider: %s, combined: %s",
+            provider_name, offer_score, provider_score, combined_score
+        )
+        return combined_score
+
+    def _final_score(self, base_score: float, provider_score: float) -> float:
+        #   NOTE: this logic is just a POC
+        if provider_score < -1.5:
+            return -1
+        return base_score + provider_score
+
+    ######################
+    #   PAYMENT MANAGEMENT
+    def on_event(self, event: events.Event):
+        if isinstance(event, events.ActivityEvent):
+            self._agreement_activity_map[event.agreement.id].add(event.activity.id)
+
+        if isinstance(event, events.WorkerFinished):
+            if event.exception is not None:
+                self._activity_failed(event, "WORKER EXCEPTION")
+        elif isinstance(event, events.TaskRejected):
+            self._activity_failed(event, "TASK REJECTED")
+
+        elif isinstance(event, events.DebitNoteAccepted):
+            activity_id = event.debit_note.activity_id
+            prev_accepted_amount = self._accepted_amounts[activity_id]
+            new_accepted_amount = max(prev_accepted_amount, Decimal(event.debit_note.total_amount_due))
+            self._accepted_amounts[activity_id] = new_accepted_amount
+            logger.warning("Accepted debit note for %s, total accepted amount: %s", activity_id, new_accepted_amount)
+
+    def _activity_failed(self, event: events.ActivityEvent, reason: str):
+        activity_id = event.activity.id
+
+        self._failed_activities.add(activity_id)
+
+        provider_name = event.provider_info.name
+        logger.warning("Disabling payments for activity %s on %s, reason %s", activity_id, provider_name, reason)
+
+    async def debit_note_accepted_amount(self, debit_note: "DebitNote") -> Decimal:
+        activity_id = debit_note.activity_id
+
+        if activity_id in self._failed_activities:
+            #   NOTE: currently it doesn't really matter what we return here,
+            #         as long as it is not debit_note.total_amount_due
+            return self._accepted_amounts[activity_id]
+        return Decimal(debit_note.total_amount_due)
+
+    async def invoice_accepted_amount(self, invoice: "Invoice") -> Decimal:
+        agreement_id = invoice.agreement_id
+        if self._agreement_has_failed_activity(agreement_id):
+            accepted_amount = self._total_agreement_amount(agreement_id)
+            logger.warning("REJECTED INVOICE FOR %s, we accept only %s", invoice.amount, accepted_amount)
```

Comment:
shouldn't we verify the amount on the invoice before we say that we reject it? however unlikely, maybe the provider will issue invoice with the accepted amount and not higher...